File: dart/glue/lib/src/lib/list/tail.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Tail function - returns the rest of a list after the first element
/// Mirrors Haskell Glue.Lib.List.Tail.tail exactly
Ir tail = IrNativeFunc(tailImpl);

/// Tail function implementation
/// Mirrors Haskell Glue.Lib.List.Tail.tailImpl exactly
Eval<Ir> tailImpl(Ir arg) {
  return switch (arg) {
    IrList(:final elements) =>
      elements.isEmpty
          ? throwError(wrongArgumentType(['non-empty list']))
          : Eval.pure(IrList(elements.skip(1).toList())),
    _ => throwError(wrongArgumentType(['list'])),
  };
}

File: haskell/glue/src/Glue/Lib/List/Tail.hs
module Glue.Lib.List.Tail where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

tail :: IR Eval
tail = NativeFunc tailImpl

tailImpl :: IR Eval -> Eval (IR Eval)
tailImpl arg = case arg of
    List (_ : xs) -> pure $ List xs
    List [] -> throwError $ wrongArgumentType ["non-empty list"]
    _ -> throwError $ wrongArgumentType ["list"]
